wireguard:overlay: cleanup calls map when unused

smagnani96 opened the pull request:
This commit cleanups the `cilium_calls_wireguard*` and `cilium_calls_overlay*`
tail call map when not used anymore. This might happen when switching
from true to false EnableWireguard/RoutingMode=Tunnel upon patch/downgrade/upgrade.
Despite being highly discourages, if someone attempts to do that, the
cilium_wg0/cilium_vxlan is being deleted as well as all the attached programs.
However, we do not cleanup the tail call map, which would stay pinned
but unused unless manually removed.

Adding `cleanCallsMaps()` utility to accomplish that. It accepts either
a specific mapName or a pattern, which is then solved through filepath.Glob.
This can be useful for those cases in which we have multiple dangling maps
referring to the same program (ex disable WireGuard during downgrade
from v1.18 to v1.17, we'd need to remove both `cilium_calls_wireguard_2`
and `cilium_calls_wireguard_<ifindex>` where we cannot retrieve ifindex
anymore). We do not check for the returned error code since this might not
be a blocking operation (at the moment of executing this the feature
Wireguard/Tunnel is disabled, so there's no reason to fail the DP
reinitialization for something that's not needed).

dylandreimerink commented on pkg/datapath/loader/base.go:
```diff
@@ -162,6 +162,20 @@ func cleanIngressQdisc(devices []string) error {
 }
 
+// cleanCallsMaps is used to remove any pinned map matching mapName from bpf.TCGlobalsPath().
+func cleanCallsMaps(mapName string) error {
```
I would rename the parameter to something like `mapNamePattern`, otherwise its unclear that glob patterns are allowed.

smagnani96 committed:
loader:wireguard:overlay: cleanup calls_map when not used

This commit cleanups the `cilium_calls_wireguard*` and `cilium_calls_overlay*`
tail call map when not used anymore. This might happen when switching
from true to false EnableWireguard/RoutingMode=Tunnel upon patch/downgrade/upgrade.
Despite being highly discourages, if someone attempts to do that, the
cilium_wg0/cilium_vxlan is being deleted as well as all the attached programs.
However, we do not cleanup the tail call map, which would stay pinned
but unused unless manually removed.

Adding `cleanCallsMaps()` utility to accomplish that. It accepts either
a specific mapName or a pattern, which is then solved through filepath.Glob.
This can be useful for those cases in which we have multiple dangling maps
referring to the same program (ex disable WireGuard during downgrade
from v1.18 to v1.17, we'd need to remove both `cilium_calls_wireguard_2`
and `cilium_calls_wireguard_<ifindex>` where we cannot retrieve ifindex
anymore). We do not check for the returned error code since this might not
be a blocking operation (at the moment of executing this the feature
Wireguard/Tunnel is disabled, so there's no reason to fail the DP
reinitialization for something that's not needed).

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

dylandreimerink reviewed (approved):
This looks good to me
